Update WebView2 Debug Visual Studio Code Docs and add UWP Debugging instructions. (#2381)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/how-to/debug-visual-studio-code.md

# Debug WebView2 apps with Visual Studio Code

Use Microsoft Visual Studio Code to debug scripts that run in WebView2 controls.  <!-- Make sure you're using Visual Studio Code version [insert build here] or later. -->
Visual Studio Code has a built-in [debugger](https://code.visualstudio.com/docs/nodejs/browser-debugging) for browser debugging.

<!-- ====================================================================== -->
## Create a launch.json file

> To debug your code, your project is required to have a `launch.json` file.  There are two request types, `launch` and `attach`. The following code demonstrates launching the app from VSCode. To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:

Reviewer: ```suggestion
To debug your code, your project is required to have a `launch.json` file.  There are two request types, `launch` and `attach`. The following code demonstrates launching the app from Visual Studio Code. To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:
```

> To debug your code, your project is required to have a `launch.json` file.  There are two request types, `launch` and `attach`. The following code demonstrates launching the app from Visual Studio Code. To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:

Reviewer: ```suggestion
To debug your code, your project is required to have a `launch.json` file.  There are two request types, `launch` and `attach`. The following code demonstrates launching the app from Visual Studio Code (rather than attaching the debugger to a running browser instance). To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:
```
or separate first part into a paragraph defining the two request types, launch vs. attach.

Reviewer 2: I find that the "there are two request types" part comes a bit out the blue. If I've never used VSCode before (or if I've used it but never came across launch.json files), I don't know what a request is. It's great to say that a launch.json file is needed, but the sentence could use something that says "A launch.json file describes how ... and needs to define a request type that describes ....". And then you can go on to say that there are two types of requests.
Does that make sense?

> To debug your code, your project is required to have a `launch.json` file.  A `launch.json` file is a debugger configuration file to configure and customize the Visual Studio Code debugger. One of the properties needed to configure the debugger is the `request` property. There are two `request` types, `launch` and `attach`. The following code demonstrates launching the app from Visual Studio Code (rather than attaching the debugger to a running instance of the app). To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:

Reviewer: ```suggestion
To debug your code, your project is required to have a `launch.json` file.  A `launch.json` file is a debugger configuration file to configure and customize the Visual Studio Code debugger. One of the properties needed to configure the debugger is the `request` property. There are two `request` types, `launch` and `attach`.

The following code demonstrates launching the app from Visual Studio Code (rather than attaching the debugger to a running instance of the app). To do this, the app must have been built previously. If your project doesn't have a `launch.json` file, create a new `launch.json` file in the `.vscode` subfolder in your current project and paste the following code into it:
```
not sure if this breaking up of paragraph was considered before.  for increments of readable text.

# Debug WebView2 apps with Visual Studio Code

Use Microsoft Visual Studio Code to debug scripts that run in WebView2 controls.  <!-- Make sure you're using Visual Studio Code version [insert build here] or later. -->

> Visual Studio Code has a built-in [debugger](https://code.visualstudio.com/docs/nodejs/browser-debugging) for browser debugging.

Reviewer: ```suggestion
Visual Studio Code has a built-in debugger for browser debugging.  See [Browser debugging in VS Code](https://code.visualstudio.com/docs/nodejs/browser-debugging).
```
per repo style guide, d/k if shown in wiki > Edge Dev Style Guide.  
And per corp-level (tho case-by-case) style guide, show actual destination heading in link text.  (in most cases, where fits)

<!-- ====================================================================== -->
## Create a launch.json file

> To debug your code, your project is required to have a `launch.json` file.  A `launch.json` file is a debugger configuration file to configure and customize the Visual Studio Code debugger. One of the properties needed to configure the debugger is the `request` property. There are two `request` types, `launch` and `attach`. 

Reviewer: ```suggestion
To debug your code, your project is required to have a `launch.json` file.  A `launch.json` file is a debugger configuration file to configure and customize the Visual Studio Code debugger. One of the properties that's needed to configure the debugger is the `request` property. There are two `request` types, `launch` and `attach`. 
```

1. To set a breakpoint in your source code, click a line of code and then press `F9`:

   ![A breakpoint that's set in Visual Studio Code](./media/breakpoint-vs.png)

1. On the **Run** tab, select the launch configuration from the dropdown menu.

> 1. Click **Start Debugging**, which is the green triangle next to the launch configuration dropdown.

Reviewer: ```suggestion
1. Click **Start Debugging**, which is the green triangle next to the launch configuration dropdown:
```

   ![The Run tab in Visual Studio Code](./media/run-vscode.png)

> 1. To view the debug output and errors, open **Debug Console**.

Reviewer: ```suggestion
1. To view the debug output and errors, open **Debug Console**:
```
colon where possible to glue flow

1. Click **Start Debugging**, which is the green triangle next to the launch configuration dropdown.

>    ![The Run tab in Visual Studio Code.](./media/run-vscode.png)

Reviewer: ```suggestion
   ![The Run tab in Visual Studio Code](./media/run-vscode.png)
```
moving away from period in alt text for simplicity and common practice, might be in wiki > Edge Dev Style Guide.  It's clear we'll need to do a global PR to stop them propagating.

"useWebView": true
```

Your WebView2 control must open the CDP port to allow debugging of the WebView2 control.  Your code must be built to ensure that only one WebView2 control has a Chrome Developer Protocol (CDP) port open, before starting the debugger. 

> You will also need to add a new REGKEY `*=--remote-debugging-port=9222` under `Computer\HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` so that the debugger can find the proper port. Open the `Registry Editor` by clicking the **Windows Key** and searching for `Registry Editor`. Launch the `Registry Editor` application and select **Yes** to allow editing. Set the registry key “HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments” equal to “--remote-debugging-port=9222”.  To do this, in the editor navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path. If this path does not exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor and right click on the `Microsoft` folder and select **New** and then select **Key**.  Enter `Edge` for the name of the new key and continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`. Right click on the `AdditionalBrowserArguments` folder and select **New** and then select **String Value**. Rename `New Value #1` to `*`. Right click on the **\*** value and select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify the edit window matches the following:

Reviewer 2: There seems to be a leading `=` sign in `*=--remote-debugging-port=9222`.

Also, consider splitting this paragraph into several lines, and possibly using a numbered list (1, 2, 3) for these steps.

"useWebView": true
```

> Your WebView2 control must open the CDP port to allow debugging of the WebView2 control.  Your code must be built to ensure that only one WebView2 control has a Chrome Developer Protocol (CDP) port open, before starting the debugger. You will also need to add a new REGKEY `*=--remote-debugging-port=9222` under `Computer\HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` so that the debugger can find the proper port.

Reviewer 2: There will be people reading this doc who have never delt with regkeys before. Can we add something here that explains where to do this? How to open the tool to add the key, etc.

1. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-port=9222`.

   To do this, in the editor, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path.

   If this path doesn't exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor, right-click the `Microsoft` folder, select **New**, and then select **Key**.  Enter `Edge` for the name of the new key.  Continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`.

1. Right-click the `AdditionalBrowserArguments` folder, select **New**, and then select **String Value**.
Rename `New Value #1` to `*`.

1. Right click the **\*** value, and then select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify that the edit window matches the following:

   ![Set Registry Key](./media/set-debugging-port.png)

> 1. Click **OK**, and then verify that the registry key is set in the editor and matches the following

Reviewer: ```suggestion
1. Click **OK**, and then verify that the registry key is set in the editor and matches the following:
```

Your WebView2 control must open the CDP port to allow debugging of the WebView2 control.  Your code must be built to ensure that only one WebView2 control has a Chrome Developer Protocol (CDP) port open, before starting the debugger.

You will also need to add a new REGKEY `*--remote-debugging-port=9222` under `Computer\HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` so that the debugger can find the proper port.

> Open the `Registry Editor` by clicking the **Windows Key** and searching for `Registry Editor`. Launch the `Registry Editor` application and select **Yes** to allow editing. Set the registry key “HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments” equal to “--remote-debugging-port=9222”.  

Reviewer: ```suggestion
Open the `Registry Editor` by clicking the **Windows Key** and searching for `Registry Editor`. Launch the `Registry Editor` application and select **Yes** to allow editing. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-port=9222`.
```
maybe try backticks (code).  When emphasizing GUI as the context, bold.  When emphasizing code as the context, backticks.

> To do this, in the editor navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path. If this path does not exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor and right click on the `Microsoft` folder and select **New** and then select **Key**.  Enter `Edge` for the name of the new key and continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`.

Reviewer: ```suggestion
To do this, in the editor, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path. 

If this path doesn't exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor and right-click the `Microsoft` folder and select **New** and then select **Key**.  Enter `Edge` for the name of the new key, and continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`.
```
corp-level style guide has a few scattered indications to say "right-click a foo" not "right-click on a foo".  eg:
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/term-collections/mouse-mouse-interaction-terms

> Right click on the `AdditionalBrowserArguments` folder and select **New** and then select **String Value**. Rename `New Value #1` to `*`. Right click on the **\*** value and select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify the edit window matches the following:

Reviewer: ```suggestion
Right click on the `AdditionalBrowserArguments` folder and select **New** and then select **String Value**. Rename `New Value #1` to `*`. Right click on the **\*** value and select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify that the edit window matches the following:
```

>    ![Set Registry Key](./media/set-debugging-port.png)

Reviewer: ```suggestion
![Set Registry Key](./media/set-debugging-port.png)
```

> Click **OK** and verify the registry key is set in the editor and matches the following:

Reviewer: ```suggestion
Click **OK**, and then verify that the registry key is set in the editor and matches the following:
```

>    ![Registry Key](./media/set-debugging-port-registry-key.png)

Reviewer: ```suggestion
![Registry Key](./media/set-debugging-port-registry-key.png)
```

Your WebView2 control must open the CDP port to allow debugging of the WebView2 control.  Your code must be built to ensure that only one WebView2 control has a Chrome Developer Protocol (CDP) port open, before starting the debugger.

> You will also need to add a new REGKEY `*--remote-debugging-port=9222` under `Computer\HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` so that the debugger can find the proper port.

Reviewer: ```suggestion
You will also need to add a new REGKEY `*--remote-debugging-port=9222` under `Computer\HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` so that the debugger can find the proper port.  To add this registry key:
```

> Open the Registry Editor by clicking the `Windows logo key` and searching for **Registry Editor**. Open the Registry Editor application and select **Yes** to allow editing. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-port=9222`.

Reviewer: ```suggestion
1. Open the Registry Editor by clicking the `Windows logo key` and searching for **Registry Editor**. Open the Registry Editor application and select **Yes** to allow editing. 

1. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-port=9222`.
```

> To do this, in the editor, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path.

Reviewer: ```suggestion
   To do this, in the editor, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` by clicking on each subfolder under the path.
```

> If this path doesn't exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor, right-click the `Microsoft` folder, select **New**, and then select **Key**.  Enter `Edge` for the name of the new key.  Continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`.

Reviewer: ```suggestion
   If this path doesn't exist, navigate to `HKEY_CURRENT_USER\Software\Policies\Microsoft` in the editor, right-click the `Microsoft` folder, select **New**, and then select **Key**.  Enter `Edge` for the name of the new key.  Continue to do this for each subfolder until you have the full path: `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments`.
```
trying to convert section to numbered step list.  In VS Code, make the indent column consistent in the section.

> Right click on the `AdditionalBrowserArguments` folder and select **New** and then select **String Value**. Rename `New Value #1` to `*`. Right click on the **\*** value and select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify that the edit window matches the following:

Reviewer: ```suggestion
1. Right-click the `AdditionalBrowserArguments` folder, select **New**, and then select **String Value**. 
 Rename `New Value #1` to `*`. 

1. Right click the **\*** value, and then select **Modify**.  Set the `Value Data` equal to `--remote-debugging-port=9222`.  Verify that the edit window matches the following:
```

> ![Set Registry Key](./media/set-debugging-port.png)

Reviewer: ```suggestion
   ![Set Registry Key](./media/set-debugging-port.png)
```

> Click **OK**, and then verify that the registry key is set in the editor and matches the following:

Reviewer: ```suggestion
1. Click **OK**, and then verify that the registry key is set in the editor and matches the following:
```

> ![Registry Key](./media/set-debugging-port-registry-key.png)

Reviewer: ```suggestion
   ![Registry Key](./media/set-debugging-port-registry-key.png)
```

```

<!-- ====================================================================== -->

> ## Debug WebView2 UWP Apps

Reviewer: ```suggestion
## Debug WebView2 WinUI 2 (UWP) Apps
```
@jasonstephen15 - nomenclature, maybe we want to name this fwk/plat. as "WinUI 2 (UWP)" rather than "UWP".

Get started with WebView2 in WinUI 2 (UWP) apps
https://learn.microsoft.com/en-us/microsoft-edge/webview2/get-started/winui2

WinUI 2 (UWP) sample app
https://learn.microsoft.com/en-us/microsoft-edge/webview2/samples/webview2_sample_uwp

1. Install a WebView2 Runtime version past `106.0.1370.34`.

1. Set the registry key “HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments” equal to “--remote-debugging-pipe”.  Verify the registry key is set in the editor and that it matches this:

>    ![Registry Key](./media/uwp-debugging-registry-key.png)

Reviewer 2: It's good to add a screenshot of the tool here, but the bullet point above should also say what the tool name is and how to open it.

> 1. Add a new configuration to your `launch.json` file.  Open the file and add the following code into `launch.json`.

Reviewer: ```suggestion
1. Add a new configuration to your `launch.json` file.  Open the file and add the following code into `launch.json`:
```
colons where possible to glue flow

   ```json
   "name": "Attach to WebView2 UWP",
   "useWebView":{
      "pipeName":"JSDebugPipe"
   }
   "request": "attach",
   "type": "msedge",
   "webRoot":"${workspaceFolder}"
   ```

1. Launch your app.

> 1. Click the Start button to attach to the process and begin debugging.

Reviewer: ```suggestion
1. Click the **Start** button to attach to the process and begin debugging.
```
check UI eg tooltip, does it say Run, or Start?  Bold UI strings.

```

<!-- ====================================================================== -->

> ## Debug WebView2 WinUI 2(UWP) Apps

Reviewer: ```suggestion
## Debug WebView2 WinUI 2 (UWP) Apps
```

1. Install a WebView2 Runtime version past `106.0.1370.34`.

> 1. Open the `Registry Editor` by clicking the **Windows Key** and searching for `Registry Editor`. Launch the `Registry Editor` application and select **Yes** to allow editing. Set the registry key “HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments” equal to “--remote-debugging-pipe”.  To do this, follow the steps outlined in the [Debug running processes](#debug-running-processes) section.

Reviewer: ```suggestion
1. Open the `Registry Editor` by clicking the **Windows Key** and searching for **Registry Editor**. Open the **Registry Editor** application and select **Yes** to allow editing. 

1. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-pipe`.  To do this, follow the steps outlined in the [Debug running processes](#debug-running-processes) section above.
```
https://learn.microsoft.com/en-us/contribute/markdown-reference#apostrophes-and-quotation-marks

> 1. Verify the registry key is set in the editor and matches the following:

Reviewer: ```suggestion
1. Verify that the registry key is set in the editor and matches the following:
```

   ![Registry Key](./media/uwp-debugging-registry-key.png)

> 1. Add a new configuration to your `launch.json` file.  Open the file and add the following code into `launch.json`:

Reviewer: ```suggestion
1. Add a new configuration to your `launch.json` file.  Open `launch.json` and add the following code:
```

```

<!-- ====================================================================== -->

> ## Debug WebView2 WinUI 2 (UWP) Apps

Reviewer: ```suggestion
## Debug WebView2 WinUI 2 (UWP) apps
```
per style guide, sentence case for headings

1. Install a WebView2 Runtime version past `106.0.1370.34`.

1. Open the Registry Editor by clicking the **Windows Key** and searching for **Registry Editor**. Open the **Registry Editor** application and select **Yes** to allow editing.

1. Set the registry key `HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments` equal to `--remote-debugging-pipe`.  To do this, follow the steps outlined in the [Debug running processes](#debug-running-processes) section above.

1. Verify that the registry key is set in the editor and matches the following:

>    ![Registry Key](./media/uwp-debugging-registry-key.png)

Reviewer: ```suggestion
   ![Setting the AdditionalBrowserArguments registry key to --remote-debugging-pipe](./media/uwp-debugging-registry-key.png)
```
Doc Build claims alt-text must be unique within a .md file
